src/GafferSceneUI/OutputBuffer.cpp

```diff
@@ -0,0 +1,493 @@
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferSceneUI/Private/OutputBuffer.h"
+
+#include "IECoreGL/ShaderLoader.h"
+
+#include "IECoreImage/DisplayDriver.h"
+
+#include "OpenEXR/ImathBoxAlgo.h"
+
+#include "boost/lexical_cast.hpp"
+
+#include <unordered_set>
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace IECoreGL;
+using namespace GafferScene;
+using namespace GafferSceneUI;
+
+//////////////////////////////////////////////////////////////////////////
+// GLSL source
+//////////////////////////////////////////////////////////////////////////
+
+namespace
+{
+
+const char *g_vertexSource = R"(
+
+#version 330 compatibility
+
+in vec2 P; // Receives unit quad
+out vec2 texCoords;
+
+void main()
+{
+	vec2 p = P * 2.0 - 1.0;
+	gl_Position = vec4( p.x, p.y, 0, 1 );
+	texCoords = P * vec2( 1, -1 ) + vec2( 0, 1 );
+}
+
+)";
+
+const char *g_fragmentSource = R"(
+
+#version 330 compatibility
+
+#include "IECoreGL/ColorAlgo.h"
+
+// Assumes texture contains sorted values.
+bool contains( usampler2D array, uint value )
+{
+	uint high = uint( textureSize( array, 0 ).x - 1 );
+	uint low = 0u;
+	while( low != high )
+	{
+		uint mid = (low + high + 1u) / 2u;
+		if( texelFetch( array, ivec2( mid, 0 ), 0 ).r > value )
+		{
+			high = mid - 1u;
+		}
+		else
+		{
+			low = mid;
+		}
+	}
+	return texelFetch( array, ivec2( low, 0 ), 0 ).r == value;
+}
+
+uniform sampler2D rgbaTexture;
+uniform sampler2D depthTexture;
+uniform usampler2D idTexture;
+// Should really be 1D, but that's not supported
+// by IECoreGL's Texture and Shader::Setup classes.
+uniform usampler2D selectionTexture;
+
+in vec2 texCoords;
+layout( location=0 ) out vec4 outColor;
+
+void main()
+{
+	outColor = texture( rgbaTexture, texCoords );
+	if( outColor.a == 0.0 )
+	{
+		discard;
+	}
+
+	/// \todo : Add OCIO support for configurable color space
+	outColor = vec4(
+		outColor.a * ieLinToSRGB( outColor.r / outColor.a ),
+		outColor.a * ieLinToSRGB( outColor.g / outColor.a ),
+		outColor.a * ieLinToSRGB( outColor.b / outColor.a ),
+		outColor.a
+	);
+
+	// Input depth is absolute in camera space (completely
+	// unrelated to clipping planes). Convert to the screen
+	// space that `GL_fragDepth` needs.
+	float depth = texture( depthTexture, texCoords ).r;
+	vec4 Pcamera = vec4( 0.0, 0.0, -depth, 1.0 );
+	vec4 Pclip = gl_ProjectionMatrix * Pcamera;
+	float ndcDepth = Pclip.z / Pclip.w;
+	gl_FragDepth = (ndcDepth + 1.0) / 2.0;
+
+	uint id = texture( idTexture, texCoords ).r;
+	outColor = mix(
+		outColor,
+		vec4( 0.466, 0.612, 0.741, 1.0 ) * outColor.a,
+		0.75 * float( contains( selectionTexture, id ) )
```

Comment:
I would have expected this to be done as a preprocess or something, rather than doing a binary search per fragment, but if it works, it works.  Have you tested if the slowdown gets noticeable when selecting very large numbers of objects?  Oh, or if you're only using 1 dimension of the texture, I guess you would just run out of selection indices if you tried to select half a particle cloud?

Reply:
Used a buffer texture in https://github.com/GafferHQ/gaffer/pull/4674/commits/b79b4eb073a9a2957e605f0b8a81060b3c6acf0c.

Reply:
Also opened https://github.com/ImageEngine/cortex/pull/1264 that will remove the need for the texture unit workaround in that commit. I'd like to get this merged without having to do a Cortex release though - will remove the workaround later.